### PR TITLE
feat: Expand backup restore MCP tools

### LINKED ISSUE #22: [Task] Expand backup and restore MCP management
Parent epic: #17

## Scope

Expand the existing tenant-wide dump/restore tools into a more complete backup and restore management surface.

Current tools: `local_ydb_dump_tenant`, `local_ydb_restore_tenant`

## Capabilities

- List available dumps under the configured dump host path.
- Support table- or directory-level dump and restore paths instead of only tenant-wide `-p .`.
- Add optional post-restore verification hooks such as `scheme describe` and bounded count queries.

## Possible implementation shape

- Add `local_ydb_list_dumps`.
- Extend `local_ydb_dump_tenant` and `local_ydb_restore_tenant` with `path`.
- Or add narrower `local_ydb_dump_path` and `local_ydb_restore_path` tools.

## Acceptance criteria

- Existing tenant-wide dump/restore behavior remains compatible.
- Path-level operations validate paths and preserve auth handling.
- Restore remains high-risk and plan-only without `confirm: true`.
- Tests cover dump listing, path-level command construction, and restore gating.
- README and MCP tool scenarios are updated.

### PR DESCRIPTION
**Pull request by astandrik**
Closes https://github.com/astandrik/local-ydb-toolkit/issues/22

<details open><summary><h3>Greptile Summary</h3></summary>

This PR expands the backup/restore MCP toolset by adding a new `local_ydb_list_dumps` read-only tool and extending `local_ydb_dump_tenant` / `local_ydb_restore_tenant` with path-level operation support, post-restore verification hooks (`describePaths`, `countQueries`), and shell-quoting fixes for dump paths that contain spaces.

- **`listDumps`** scans `profile.dumpHostPath` on the host via a bash spec, filtering to top-level directories that contain a `tenant` subfolder and no `incomplete` marker; results are parsed through `normalizeListedDumpName` which enforces a strict round-trip check to keep only safe single-component directory names.
- **`dumpTenant` / `restoreTenant`** gain a `path` field wired through `normalizeYdbRelativePath` (rejects absolute paths, `..`, empty components), and `restoreTenant` appends optional `schemeDescribe` and `countQuery` hooks validated by `normalizeCountQuery` (regex-gated to `SELECT COUNT(*|1) FROM \`relative/path\`` plus UNION/INTERSECT/EXCEPT rejection via `stripQuotedSqlText`).
- All changes are fully covered by unit tests (mock-executor and real-filesystem variants) and a new end-to-end `verifyBackupRestore` CI scenario that exercises plan-only → confirmed dump → list → plan-only restore → confirmed restore with count verification.
</details>

<h3>Confidence Score: 5/5</h3>

Safe to merge; changes are additive, well-validated, and backward-compatible with existing tenant-wide dump/restore calls.

All new code paths have strong input validation at both the Zod/MCP layer and the core layer, rollback/verification paths are now properly shell-quoted, and the test suite covers edge cases including shell-injection guard tokens in backtick-quoted identifiers, set-operation rejection, byte-limit enforcement for multibyte characters, and real-filesystem incomplete-marker filtering.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| packages/core/src/operations/tenant.ts | Core implementation of listDumps, expanded dumpTenant (with path support and shellQuote fixes), and restoreTenant (with path, describePaths, countQueries verification hooks); validation helpers are thorough and well-tested. |
| packages/core/src/operations/types.ts | Adds DumpTenantOptions, RestoreTenantOptions, DumpTenantResponse, RestoreTenantResponse, ListDumpsResponse, and RestoreVerificationHook types; optional fields on RestoreTenantResponse correctly accommodate the early-return no-dumpName path. |
| packages/mcp-server/src/tools/args.ts | DumpArgs gains path and tightens dumpName to trim+min(1); RestoreArgs adds path, describePaths, and countQueries with a RestoreCountQueryArgs sub-schema that byte-checks the query using Buffer.byteLength matching core's limit. |
| packages/mcp-server/src/tools/registry.ts | Registers local_ydb_list_dumps as a read-only tool using ProfileArgs; updates dump and restore descriptions to reflect new path and hook capabilities. |
| scripts/ci/verify-live-mcp-server.mjs | Adds an end-to-end verifyBackupRestore CI flow covering path-level dump, list-dumps, path-level restore, and countQuery verification; authentication is handled correctly via docker exec with a temp password file, and cleanup is attempted in finally with proper error propagation. |
| packages/core/test/operations.test.ts | Adds comprehensive unit tests for listDumps (mocked and real-fs), path-level dump/restore, verification hooks, set-operation rejection, and count-query guard tokens in quoted identifiers. |
| packages/mcp-server/src/tools/input-schemas.ts | Adds listDumpsSchema and extends dumpSchema/restoreSchema with path, describePaths, and countQueries; additionalProperties: false is correctly set on all objects including the countQuery item schema. |
| packages/mcp-server/test/tools.test.ts | Adds MCP-layer handler tests for list dumps, path-level dump, blank-arg rejection, restore verification hooks, byte-limit rejection, and blank label/path/describePaths/countQuery rejection. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[MCP Client] -->|local_ydb_list_dumps| B[ListDumpsArgs/ProfileArgs]
    A -->|local_ydb_dump_tenant| C[DumpArgs<br/>dumpName? path?]
    A -->|local_ydb_restore_tenant| D[RestoreArgs<br/>dumpName path?<br/>describePaths? countQueries?]

    B --> E[listDumps core]
    C --> F[dumpTenant core]
    D --> G[restoreTenant core]

    E --> H[bash: list dumpHostPath/*<br/>filter tenant dir + incomplete marker]
    H --> I[parseDumpNames<br/>normalizeListedDumpName]
    I --> J[ListDumpsResponse]

    F --> K{normalizeDumpName<br/>normalizeYdbRelativePath}
    K --> L[mkdir dumpPath<br/>helperContainer ydb tools dump -p path]
    L --> M[DumpTenantResponse<br/>dumpName path sourcePath dumpPath]

    G --> N{dumpName present?}
    N -->|no| O[planOnly early return<br/>verificationHooks: empty]
    N -->|yes| P[restoreVerificationHooks<br/>normalizeCountQuery]
    P --> Q[helperContainer ydb tools restore<br/>+ schemeDescribe hooks<br/>+ countQuery hooks]
    Q --> R[RestoreTenantResponse<br/>dumpName path targetPath verificationHooks]
```
</details>

<sub>Reviews (10): Last reviewed commit: ["Allow semicolons in quoted count paths"](https://github.com/astandrik/local-ydb-toolkit/commit/2372322bc6009fb1eb961df4a6e0f4a844efb3b3) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=37305418)</sub>

### CODE REVIEW

File: packages/core/src/operations/tenant.ts [hunk 1]
```diff
@@ -15,33 +26,196 @@ export async function createTenant(ctx: ToolkitContext, options: MutatingOptions
...
+    helperContainer(ctx.profile, `/ydb -e grpc://localhost:${ctx.profile.ports.dynamicGrpc} -d ${shellQuote(ctx.profile.tenantPath)} ${ydbAuthArgs(ctx.profile)} tools dump -p ${shellQuote(path)} --exclude ${shellQuote(SYSTEM_PATH_EXCLUDE_PATTERN)} -o ${shellQuote(`/dump/${dumpName}/tenant`)}`)
   ];
-  return runMutating(ctx, {
-    summary: `Dump ${ctx.profile.tenantPath} to ${dumpPath}.`,
+  const response = await runMutating(ctx, {
+    summary: `Dump ${sourcePath} to ${dumpPath}.`,
     risk: "medium",
     specs,
     rollback: [`rm -rf ${dumpPath}`],
     verification: [`test -d ${dumpPath}/tenant`]
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/tenant.ts]: Fixed in 5ff40da: dump rollback and verification snippets now quote dump paths with shellQuote, including a regression case for a dumpName containing a space.

File: skills/local-ydb/references/mcp-tool-scenarios.md [hunk 2]
```diff
@@ -398,14 +399,25 @@ Calls:
...
 { "tool": "local_ydb_tenant_check", "arguments": { "profile": "ghcr261-clean" } }
 { "tool": "local_ydb_graphshard_check", "arguments": { "profile": "ghcr261-clean" } }
 ```
 
+Path-level example:
+
+```json
+{ "tool": "local_ydb_dump_tenant", "arguments": { "profile": "local", "dumpName": "one-table-smoke", "path": "dir/table" } }
+{ "tool": "local_ydb_restore_tenant", "arguments": { "profile": "ghcr261-clean", "dumpName": "one-table-smoke", "path": ".", "describePaths": ["dir/table"], "countQueries": [{ "label": "dir/table rows", "query": "SELECT COUNT(*) FROM `dir/table`;" }] } }
+```
```
[the remark this answers is not in the document]
Comment (astandrik) [reply skills/local-ydb/references/mcp-tool-scenarios.md:413]: Fixed in 5ff40da: the path-level dump/restore example now includes confirm:true on both calls, so it matches the scenario goal and runs restore hooks.

File: MCP_TOOL_TEST_SCENARIOS.md [hunk 3]
```diff
@@ -442,14 +443,25 @@ Calls:
...
 { "tool": "local_ydb_tenant_check", "arguments": { "profile": "ghcr261-clean" } }
 { "tool": "local_ydb_graphshard_check", "arguments": { "profile": "ghcr261-clean" } }
 ```
 
+Path-level example:
+
+```json
+{ "tool": "local_ydb_dump_tenant", "arguments": { "profile": "local", "dumpName": "one-table-smoke", "path": "dir/table" } }
+{ "tool": "local_ydb_restore_tenant", "arguments": { "profile": "ghcr261-clean", "dumpName": "one-table-smoke", "path": ".", "describePaths": ["dir/table"], "countQueries": [{ "label": "dir/table rows", "query": "SELECT COUNT(*) FROM `dir/table`;" }] } }
+```
```
[the remark this answers is not in the document]
Comment (astandrik) [reply MCP_TOOL_TEST_SCENARIOS.md:457]: Fixed in 5ff40da: the path-level dump/restore example now includes confirm:true on both calls, so it creates the dump and executes restore verification hooks.

File: packages/mcp-server/test/tools.test.ts [hunk 4]
```diff
@@ -543,6 +546,58 @@ describe("mcp tools", () => {
...
+    executor.run = async (_profile, spec) => {
+      const command = executor.display(_profile, spec);
+      executor.commands.push(command);
+      return { command, exitCode: 0, stdout: "mcp-smoke\n", stderr: "", ok: true, timedOut: false };
+    };
+
+    const result = await callLocalYdbToolForTest("local_ydb_list_dumps", {}, {
+      config: ConfigSchema.parse({}),
+      executor
+    }) as { ok: boolean; dumps: Array<{ name: string; tenantDumpPath: string }> };
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/mcp-server/test/tools.test.ts]: Fixed in 5ff40da: the handler test assertion type now includes hostPath alongside name and tenantDumpPath.

File: packages/core/src/operations/tenant.ts [hunk 5]
```diff
@@ -15,33 +26,196 @@ export async function createTenant(ctx: ToolkitContext, options: MutatingOptions
...
+
+function normalizeCountQuery(value: string): string {
+  const query = value.trim();
+  if (!query || Buffer.byteLength(query, "utf8") > MAX_COUNT_QUERY_BYTES) {
+    throw new Error(`countQueries[].query must be non-empty and at most ${MAX_COUNT_QUERY_BYTES} bytes`);
+  }
+  const withoutTrailingSemicolon = query.endsWith(";") ? query.slice(0, -1) : query;
+  if (withoutTrailingSemicolon.includes(";")) {
+    throw new Error("countQueries[].query must contain a single SELECT COUNT statement");
+  }
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/tenant.ts]: Fixed in 5ff40da: count query validation now rejects UNION in addition to semicolon-separated multi-statement input, with regression coverage.

File: packages/mcp-server/src/tools/args.ts [hunk 6]
```diff
@@ -535,10 +535,19 @@ export const DestroyStackArgs = MutatingArgs.extend({
 
 export const DumpArgs = MutatingArgs.extend({
   dumpName: z.string().optional(),
+  path: z.string().optional(),
 });
 
+const RestoreCountQueryArgs = z.object({
+  label: z.string().min(1).optional(),
+  query: z.string().min(1).max(4096),
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/mcp-server/src/tools/args.ts]: Fixed in 5ff40da: the MCP Zod layer now enforces the same UTF-8 byte limit as core, the JSON schema text documents the byte contract, and regression coverage validates multibyte rejection.

File: packages/core/src/operations/tenant.ts [hunk 7]
```diff
@@ -15,33 +26,196 @@ export async function createTenant(ctx: ToolkitContext, options: MutatingOptions
...
+    description: "List local-ydb dumps"
+  });
+  const result = await ctx.client.run(spec);
+  const dumps = result.ok
+    ? parseDumpNames(result.stdout).map((name) => ({
+        name,
+        hostPath: `${dumpHostPath}/${name}`,
+        tenantDumpPath: `${dumpHostPath}/${name}/tenant`
+      }))
+    : [];
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/tenant.ts:51]: Fixed in 259966e: local_ydb_list_dumps now filters basename output through normalizeDumpName via normalizeListedDumpName, so names that dump/restore would reject are not advertised. Regression coverage includes ../, ., .., backslashes, control characters, and trim-mismatch cases.

File: packages/core/src/operations/tenant.ts [hunk 8]
```diff
@@ -15,33 +26,196 @@ export async function createTenant(ctx: ToolkitContext, options: MutatingOptions
...
+
+function normalizeCountQuery(value: string): string {
+  const query = value.trim();
+  if (!query || Buffer.byteLength(query, "utf8") > MAX_COUNT_QUERY_BYTES) {
+    throw new Error(`countQueries[].query must be non-empty and at most ${MAX_COUNT_QUERY_BYTES} bytes`);
+  }
+  const withoutTrailingSemicolon = query.endsWith(";") ? query.slice(0, -1) : query;
+  if (withoutTrailingSemicolon.includes(";") || /\bunion\b/i.test(withoutTrailingSemicolon)) {
+    throw new Error("countQueries[].query must contain a single SELECT COUNT statement");
+  }
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/tenant.ts:235]: Fixed in 259966e: count-query validation now rejects UNION, INTERSECT, and EXCEPT after stripping quoted SQL text, and the restore hook only accepts whole-table COUNT(*)/COUNT(1) queries over a quoted relative path. Regression coverage also keeps a table named `except` valid.

File: packages/core/src/operations/tenant.ts [hunk 9]
```diff
@@ -15,33 +27,224 @@ export async function createTenant(ctx: ToolkitContext, options: MutatingOptions
...
 }
 
-export async function dumpTenant(ctx: ToolkitContext, options: MutatingOptions & { dumpName?: string } = {}) {
-  const dumpName = options.dumpName ?? `${ctx.profile.tenantPath.split("/").pop() ?? "tenant"}-${new Date().toISOString().replace(/[:.]/g, "-")}`;
+export async function listDumps(ctx: ToolkitContext): Promise<ListDumpsResponse> {
+  const dumpHostPath = ctx.profile.dumpHostPath;
+  const spec = bash([
+    `if [ -d ${shellQuote(dumpHostPath)} ]; then`,
+    `  for dir in ${shellQuote(dumpHostPath)}/*; do`,
+    "    [ -d \"$dir/tenant\" ] && basename \"$dir\"",
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/tenant.ts]: Fixed locally and verified: local_ydb_list_dumps now skips dump entries when find "$dir/tenant" -name incomplete finds a root or nested incomplete marker, so restore inventory only advertises dumps that upstream tools restore will not reject for that marker. Added a core regression test with complete, root-incomplete, nested-incomplete, and no-tenant directories. Verification: focused operations/tools tests, full npm test, npm run typecheck, npm run build, and git diff --check.

File: packages/mcp-server/src/tools/args.ts [hunk 10]
```diff
@@ -535,10 +535,22 @@ export const DestroyStackArgs = MutatingArgs.extend({
...
+  label: z.string().min(1).optional(),
+  query: z.string().min(1).refine(
+    (query) => Buffer.byteLength(query, "utf8") <= 4096,
+    { message: "query must be at most 4096 bytes" },
+  ),
+}).strict();
+
 export const RestoreArgs = MutatingArgs.extend({
-  dumpName: z.string(),
+  dumpName: z.string().min(1),
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/mcp-server/src/tools/args.ts]: Fixed locally and verified: RestoreArgs.dumpName now uses trim().min(1), so whitespace-only dumpName values are rejected at the public MCP argument layer with the same early validation path as empty strings. Added regression coverage for dumpName: "   ". Verification: focused MCP tools test, full npm test, npm run typecheck, npm run build, and git diff --check.

File: packages/mcp-server/src/tools/args.ts [hunk 11]
```diff
@@ -535,10 +535,22 @@ export const DestroyStackArgs = MutatingArgs.extend({
...
+  path: z.string().optional(),
 });
 
+const RestoreCountQueryArgs = z.object({
+  label: z.string().min(1).optional(),
+  query: z.string().min(1).refine(
+    (query) => Buffer.byteLength(query, "utf8") <= 4096,
+    { message: "query must be at most 4096 bytes" },
+  ),
+}).strict();
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/mcp-server/src/tools/args.ts:547]: Fixed locally and verified: RestoreCountQueryArgs.query now uses trim().min(1) before the UTF-8 byte-limit refine, so whitespace-only restore count queries fail in the MCP argument layer. Added regression coverage for query: "   ". Verification: focused MCP tools test, full npm test, npm run typecheck, npm run build, and git diff --check.

File: packages/mcp-server/src/tools/args.ts [hunk 12]
```diff
@@ -535,10 +535,22 @@ export const DestroyStackArgs = MutatingArgs.extend({
 
 export const DumpArgs = MutatingArgs.extend({
   dumpName: z.string().optional(),
+  path: z.string().optional(),
 });
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/mcp-server/src/tools/args.ts:539]: Fixed locally and verified: DumpArgs.dumpName and DumpArgs.path now use trim().min(1).optional(), so provided blank values are rejected in the MCP argument layer instead of falling through to core normalization. Added regression coverage for empty and whitespace-only dumpName/path. Verification: focused MCP tools test, full npm test, npm run typecheck, npm run build, and git diff --check.

File: packages/mcp-server/src/tools/args.ts [hunk 13]
```diff
@@ -535,10 +535,22 @@ export const DestroyStackArgs = MutatingArgs.extend({
...
 
+const RestoreCountQueryArgs = z.object({
+  label: z.string().min(1).optional(),
+  query: z.string().trim().min(1).refine(
+    (query) => Buffer.byteLength(query, "utf8") <= 4096,
+    { message: "query must be at most 4096 bytes" },
+  ),
+}).strict();
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/mcp-server/src/tools/args.ts:547]: Fixed locally and verified: RestoreCountQueryArgs.label now uses trim().min(1).optional(), so whitespace-only labels fail validation at the MCP argument layer instead of being silently replaced by the default label downstream. Added regression coverage for label: "   ". Verification: focused MCP tools test, full npm test, npm run typecheck, npm run build, and git diff --check.

File: packages/mcp-server/src/tools/args.ts [hunk 14]
```diff
@@ -534,11 +534,23 @@ export const DestroyStackArgs = MutatingArgs.extend({
...
+
 export const RestoreArgs = MutatingArgs.extend({
-  dumpName: z.string(),
+  dumpName: z.string().trim().min(1),
+  path: z.string().optional(),
+  describePaths: z.array(z.string().min(1)).optional(),
+  countQueries: z.array(RestoreCountQueryArgs).optional(),
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/mcp-server/src/tools/args.ts:553]: Fixed locally and verified: RestoreArgs.path and each describePaths[] item now use trim().min(1), so whitespace-only restore paths are rejected at the MCP argument layer before core normalization. Added regression coverage for path: "   " and describePaths: ["   "]. Verification: focused MCP tools test, full npm test, npm run typecheck, npm run build, and git diff --check.

File: packages/core/src/operations/tenant.ts [hunk 15]
```diff
@@ -15,33 +27,226 @@ export async function createTenant(ctx: ToolkitContext, options: MutatingOptions
...
+  if (!query || Buffer.byteLength(query, "utf8") > MAX_COUNT_QUERY_BYTES) {
+    throw new Error(`countQueries[].query must be non-empty and at most ${MAX_COUNT_QUERY_BYTES} bytes`);
+  }
+  const withoutTrailingSemicolon = query.endsWith(";") ? query.slice(0, -1) : query;
+  if (
+    withoutTrailingSemicolon.includes(";") ||
+    /\b(?:union|intersect|except)\b/i.test(stripQuotedSqlText(withoutTrailingSemicolon))
+  ) {
+    throw new Error("countQueries[].query must contain a single SELECT COUNT statement");
+  }
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/tenant.ts:235]: Fixed locally and verified: normalizeCountQuery now strips quoted SQL text once and uses that sanitized text for both the semicolon multi-statement guard and set-operation keyword guard. This keeps real multi-statement queries rejected while allowing semicolons inside quoted YDB paths such as `smoke_dst;semi/items`. Added regression coverage for the quoted-path case. Verification: npm test -- packages/core/test/operations.test.ts, npm test, npm run typecheck, npm run build, git diff --check, plus a disposable live local-ydb smoke using path-level dump/list/restore with restore path `smoke_dst;semi`, scheme describe, and a count hook returning 2 rows.
